src/lib.rs

```diff
@@ -16,116 +16,261 @@
+        let desc: protobuf::descriptor::FileDescriptorSet =
+            protobuf::parse_from_bytes(&desc_bytes).unwrap();
+        let mut files_to_generate = Vec::new();
+        'outer: for file in files {
+            for include in includes {
+                if let Ok(truncated) = file.as_ref().strip_prefix(include) {
+                    files_to_generate.push(format!("{}", truncated.display()));
+                    continue 'outer;
+                }
+            }
+
+            panic!("file {:?} is not found in includes {:?}", file, includes);
+        }
+
+        protobuf_codegen::gen_and_write(
+            desc.get_file(),
+            &files_to_generate,
+            &Path::new(out_dir),
+            &protobuf_codegen::Customize::default(),
+        )
+        .unwrap();
+        generate_grpcio(&desc.get_file(), &files_to_generate, out_dir);
+    }
+
+    #[cfg(feature = "grpcio-protobuf-codec")]
+    pub fn generate_grpcio(
+        desc: &[protobuf::descriptor::FileDescriptorProto],
+        files_to_generate: &[String],
+        out_dir: &str,
+    ) {
+        use std::io::Write;
+
+        let output_dir = std::path::Path::new(out_dir);
+        let results = grpcio_compiler::codegen::gen(desc, &files_to_generate);
+        for res in results {
+            let out_file = output_dir.join(&res.name);
+            let mut f = std::fs::File::create(&out_file).unwrap();
+            f.write_all(&res.content).unwrap();
+        }
+    }
+
+    #[cfg(all(feature = "protobuf-codec", not(feature = "grpcio-protobuf-codec")))]
+    pub fn generate_grpcio(_: &[protobuf::descriptor::FileDescriptorProto], _: &[String], _: &str) {
+    }
 }
 
-/// Use Prost to generate Rust files from proto files (`file_names`).
-///
-/// Uses `["proto", "include"]` as the include lists.
-///
-/// Returns a list of the package names of the protocols that were compiled.
-pub fn generate_prost_files<T: AsRef<str>>(file_names: &[T], out_dir: &str) -> Vec<String> {
-    let packages = grpcio_compiler::prost_codegen::compile_protos(
-        &file_names.iter().map(AsRef::as_ref).collect::<Vec<_>>(),
-        &["proto", "include"],
+#[cfg(feature = "protobuf-codec")]
+pub use protobuf_imps::*;
+
+/// Use prost to generate Rust files from proto files (`files`).
+#[cfg(all(feature = "prost-codec", not(feature = "grpcio-prost-codec")))]
+pub fn generate<T: AsRef<Path>>(includes: &[T], files: &[T], out_dir: &str) {
+    prost_build::Config::new()
+        .out_dir(out_dir)
+        .compile_protos(files, includes)
+        .unwrap();
+    let mod_names = module_names_for_dir(out_dir);
+    generate_wrappers(
+        &mod_names
+            .iter()
+            .map(|m| format!("{}/{}.rs", out_dir, m))
+            .collect::<Vec<_>>(),
         out_dir,
-    )
-    .unwrap();
+        GenOpt::MUT
+            | GenOpt::HAS
+            | GenOpt::TAKE
+            | GenOpt::CLEAR
+            | GenOpt::MESSAGE
+            | GenOpt::TRIVIAL_SET,
```

Comment:
These options should be exposed to clients, not hard-wired

Reply:
If user want to provide custom flags, they can do it via other functions. This function is just to provide a convenient assumptions for the client. In the future, we should use builder instead of pure functions.

Reply:
Then we should use `GenOpt::all()` here